compiler/vm/packed_env.nim

```diff
@@ -0,0 +1,945 @@
+                           kind: x.kind)
+    case x.kind
+    of ckDefault:
+      f.start = x.a.int
+      f.regCount = x.b.uint16
+    of ckCallback:
+      f.cbOffset = x.a.int
+
+    f
+
+  mapList(dst.debug, src.debug, x): src.infos[x.LitId]
+
+  dst.rtti = loadTypeInfos(src, dst.types)
+
+  dst.config.m.fileInfos.newSeq(src.files.len)
+  for i, f in src.files.pairs:
+    dst.config.m.fileInfos[i].fullPath = f.AbsoluteFile
+
+func init*(enc: var PackedEncoder, types: seq[PVmType]) =
+  ## Initializes the encoder. The sequence pass to `types` should not change
+  ## past this point. Else, encoding error may happen
+
+  # set up the `PVmType` -> `VmTypeId` mappings
+  # XXX: this will become unnecessary once `PVmType` is replaced by `VmTypeId`
+  for i in 0..<types.len:
+    let t = types[i]
+    enc.typeMap[t] = i.uint32
+
+  assert enc.typeMap.len == types.len
+
+func storeEnv*(enc: var PackedEncoder, dst: var PackedEnv, c: TCtx) =
+  ## Stores all relevant data provided by `c` into `dst`. Previously stored
+  ## data (except `nodes`, `numbers`, and `strings`) is thrown away. The only
+  ## parts of `dst` not touched by `storeEnv` are: `cconsts`, `globals`,
+  ## `code`, and `entryPoint`. These have to be filled in separately.
+  # store all types:
+  dst.types.newSeq(c.types.len - 1) # -1 since the invalid type is not included
+  for i in 1..<c.types.len:
+    let t = c.types[i]
+    dst.types[i - 1] = storeVmType(enc, dst, t)
+
+  # TODO: maybe make `dst.consts` a seq of indices into `dst.nodes` instead?
+  mapList(dst.consts, c.constants, c):
+    let item =
+      case c.kind
+      of cnstInt:     dst.getLitId(c.intVal).uint32
+      of cnstFloat:   dst.getLitId(c.floatVal).uint32
+      of cnstString:  dst.getLitId(c.strVal).uint32
+
+      of cnstSliceListInt:   dst.storeSliceList(c.intSlices)
+      of cnstSliceListFloat: dst.storeSliceList(c.floatSlices)
+      of cnstSliceListStr:   dst.storeSliceList(c.strSlices)
+      of cnstNode:
+        # node constants are not created in a non-compiletime context
+        unreachable()
+
+    (c.kind, item)
+
+  mapList(dst.functions, c.functions, x):
+    let
+      d = dst.storeDbgSym(x.sym)
+      t1 = enc.typeMap[x.retValDesc]
+      t2 = enc.typeMap[x.envParamType]
+
+    let (a, b) =
+      case x.kind
+      of ckCallback: (x.cbOffset.uint32, 0'u32)
+      of ckDefault:  (x.start.uint32, x.regCount.uint32)
+
+    (d, x.sig, t1, t2, x.kind, a, b)
+
+  mapList(dst.debug, c.debug, d):
+    dst.infos.getOrIncl(d).uint32
+
+  # TODO: add support for either `distinct string` or custom `storePrim`
+  #       overloads (or both) to `rodfiles`. Due to the lack of both, we
+  #       have to perform a manual copy instead of a move here
+  mapList(dst.callbacks, c.callbackKeys, c):
+    c.string
+
+  mapList(dst.files, c.config.m.fileInfos, fi):
+    fi.fullPath.string
+
+  storeTypeInfos(enc, dst, c.rtti)
+
+proc writeToFile*(p: PackedEnv, file: AbsoluteFile): RodFileError =
```

Comment:
Tests and/or template might help here to ensure the symmetry.

Reply:
Added a test (`tests/compiler/tvmbackend.nim`).